### DISCUSSION TITLE
"vehicle_data_combo" endpoint of "vehicle_data" command

### DISCUSSION THREAD

**Discussion by GaPhi** [timdorr/tesla-api#655] [General]
Hello,

Did someone has found how to decode the new "vehicle_data_combo" endpoint of "vehicle_data" command ?
Data seem to be in a more efficient form.
It returns a Base64 encoded buffer into "vehicle_data" field, and is not a BSON format.

After Base64 decoding, we can find some strings into the buffer prefixed by its length as a UINT8 byte.
But software version is a bit different with a MSB set to 1 : the string length prefix is 0x98 for a 24 chars string.

I also seen that 0xC2 is frequent (0b11000011 byte) and 0xC3, ... which look like some kind of marker/starter.

Any other has already investigated it ?

**Comment by strawhatguy** [comment]
Were you able to figure out what this was?  I'm thinking it is a protobuf version of the data, like what comes in via their fleet-telemetry websocket stream per vehicle: https://github.com/teslamotors/fleet-telemetry/blob/main/protos/vehicle_data.pb.go#L969

However, I tried to unmarshal it, and got a field string that wasn't UTF-8, so the encoding failed.

**Reply by flarno11** [reply]
It's not the same definition as the one you linked. But `vehicle_data_combo` and `cached_data` from the `/api/1/products` endpoint look similar and some parts are even identical.
[protobuf-inspector](https://github.com/mildsunrise/protobuf-inspector) helps investigating and comparing the data.
It looks like it is a direct copy of the `vehicle_data` json structure, except `cached_data` contains additional 2x 512 bytes blocks at the end.

I tried to extract proto files from the Android apk using https://github.com/marin-m/pbtk but it only found `energy.proto`/ `tesla.proto.charging.v1`, nothing related to the `vehicle_data`.
